haystack/utils/device.py

```diff
@@ -1,34 +1,471 @@
+    def resolve_device(device: Optional["ComponentDevice"] = None) -> "ComponentDevice":
+        """
+        Select a device for a component. If a device is specified,
+        it's used. Otherwise, the default device is used.
+
+        :param device:
+            The provided device, if any.
+        :returns:
+            The resolved device.
+        """
+        if not isinstance(device, ComponentDevice) and device is not None:
+            raise ValueError(
+                f"Invalid component device type '{type(device).__name__}'. Must either be None or ComponentDevice."
+            )
+
+        if device is None:
+            device = ComponentDevice.from_single(_get_default_device())
+
+        return device
+
+    def to_dict(self) -> Dict[str, Any]:
+        """
+        Convert the component device representation to a JSON-serializable
+        dictionary.
+
+        :returns:
+            The dictionary representation.
+        """
+        if self._single_device is not None:
+            return {"type": "single", "device": str(self._single_device)}
+        elif self._multiple_devices is not None:
+            return {"type": "multiple", "device_map": self._multiple_devices.to_dict()}
+        else:
+            # Unreachable
+            assert False
+
+    @classmethod
+    def from_dict(cls, dict: Dict[str, Any]) -> "ComponentDevice":
+        """
+        Create a component device representation from a JSON-serialized
+        dictionary.
+
+        :param dict:
+            The serialized representation.
+        :returns:
+            The deserialized component device.
+        """
+        if dict["type"] == "single":
+            return cls.from_str(dict["device"])
+        elif dict["type"] == "multiple":
+            return cls.from_multiple(DeviceMap.from_dict(dict["device_map"]))
+        else:
+            raise ValueError(f"Unknown component device type '{dict['type']}' in serialized data")
+
+
+def _get_default_device() -> Device:
+    """
+    Return the default device for Haystack. Precedence:
+    GPU > MPS > CPU. If PyTorch is not installed, only CPU
+    is available.
+
+    :returns:
+        The default device.
     """
     try:
         torch_import.check()
-    except ImportError as e:
-        logger.warning(e.msg)
-        return "cpu:0"
-
-    if torch.cuda.is_available():
-        device = "cuda:0"
-    elif (
-        hasattr(torch.backends, "mps")
-        and torch.backends.mps.is_available()
-        and os.getenv("HAYSTACK_MPS_ENABLED", "true") != "false"
-    ):
-        device = "mps:0"
+        has_torch = True
+    except ImportError:
+        has_torch = False
+
+    if not has_torch:
+        has_mps = False
+        has_cuda = False
+    else:
+        has_mps = (
+            hasattr(torch.backends, "mps")
+            and torch.backends.mps.is_available()
+            and os.getenv("HAYSTACK_MPS_ENABLED", "true") != "false"
+        )
+        has_cuda = torch.cuda.is_available()
```

Comment:
You could merge these two:

```suggestion
        has_mps = (
            hasattr(torch.backends, "mps")
            and torch.backends.mps.is_available()
            and os.getenv("HAYSTACK_MPS_ENABLED", "true") != "false"
        )
        has_cuda = torch.cuda.is_available()
    except ImportError:
        has_mps = False
        has_cuda = False
```
but it's not necessary, feel free to discard it if you prefer the two separate checks.

Reply:
Merged.